hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java

```diff
@@ -1753,7 +1756,28 @@ public ConnectionId(InetSocketAddress address, Class<?> protocol,
     InetSocketAddress getAddress() {
       return address;
     }
-    
+
+    /**
+     * This is used to update the remote address when an address change is detected.  This method
+     * ensures that the {@link #hashCode()} won't change.
+     *
+     * @param address the updated address
+     * @throws IllegalArgumentException if the hostname or port doesn't match
+     * @see Connection#updateAddress()
+     */
+    void setAddress(InetSocketAddress address) {
+      if (!Objects.equals(this.address.getHostName(), address.getHostName())) {
+        throw new IllegalArgumentException("Hostname must match: " + this.address + " vs "
+            + address);
+      }
+      if (this.address.getPort() != address.getPort()) {
+        throw new IllegalArgumentException("Port must match: " + this.address + " vs " + address);
+      }
+
+      this.address = address;
```

Comment:
Should we add some log saying the Address is changed from XXX to XXX?

Reply:
There is a message in the Connection#updateAddress call.  This services that call to update the remoteId to avoid having to re-resolve the updated IP address in subsequent calls.  The look like the following:
```
namenode-0 namenode 2022-08-08 13:43:02 WARN  Client:639 - Address change detected. Old: journalnode-0.journalnode.hdfs.svc.cluster.local/10.1.6.87:8485 New: journalnode-0.journalnode.hdfs.svc.cluster.local/10.1.6.88:8485
```